### PR TITLE
Update the Adam optimizer class API docs

### PR DESCRIPTION
**Pull request by 8bitmp3**
This PR updates the Adam optimizer class API docs, as discussed with @saeta. API docs are important and a few members of the OS community have already looked at things like Sequential and losses.

Introducing a few things to assist the s4tf users:

- Expanded the description of Adam to be closer to the one in the TF 2.x docs (thanks @lamberta and the docs team).
- Defined parameters.
- Added two examples: RL and GAN code snippets adapted from `/swift-models/` (Gym and GAN examples).

I'm going through other commonly used classes, funcs, structs and protocols (thought Adam would be a good start), so will be sending more PRs. 

Your comments, suggestions, bug fixes are most welcome. 

Thanks for all the hard work ☕️

### CODE REVIEW

**Comment by saeta** [comment]
Thank you @8bitmp3 !!

File: Sources/TensorFlow/Optimizers/MomentumBased.swift [hunk 1]
```diff
@@ -162,17 +162,101 @@ public class AdaDelta<Model: Differentiable>: Optimizer
...
+///         // with the Adam optimizer.
+///         adamOptimizerG.update(&generator, along: 𝛁generator)
+///
+///         // 2) Update the discriminator.
+///         ...
+///         let discriminator = TensorFlow.gradient(at: generator) { generator -> Tensor<Float> in
+///             ...
+///             return loss
+///         }
+///         // Update the differentiable variables of the generator along the gradients (`𝛁discriminator`)
```
Comment (t-ae) [inline Sources/TensorFlow/Optimizers/MomentumBased.swift]: Update the differentiable variables of the **discriminator**

File: Sources/TensorFlow/Optimizers/MomentumBased.swift [hunk 2]
```diff
@@ -162,17 +162,101 @@ public class AdaDelta<Model: Differentiable>: Optimizer
...
+///
+///         // 2) Update the discriminator.
+///         ...
+///         let discriminator = TensorFlow.gradient(at: generator) { generator -> Tensor<Float> in
+///             ...
+///             return loss
+///         }
+///         // Update the differentiable variables of the generator along the gradients (`𝛁discriminator`) 
+///         // with the Adam optimizer.
+///         adamOptimizerD.update(&generator, along: 𝛁discriminator)
```
Comment (t-ae) [inline Sources/TensorFlow/Optimizers/MomentumBased.swift]: adamOptimizerD.update(&**discriminator**, along: 𝛁discriminator)

File: Sources/TensorFlow/Optimizers/MomentumBased.swift [hunk 3]
```diff
@@ -162,17 +162,101 @@ public class AdaDelta<Model: Differentiable>: Optimizer
...
+///             ...
+///             return loss
+///             }
+///         // Update the differentiable variables of the generator along the gradients (`𝛁generator`) 
+///         // with the Adam optimizer.
+///         adamOptimizerG.update(&generator, along: 𝛁generator)
+///
+///         // 2) Update the discriminator.
+///         ...
+///         let discriminator = TensorFlow.gradient(at: generator) { generator -> Tensor<Float> in
```
Comment (t-ae) [inline Sources/TensorFlow/Optimizers/MomentumBased.swift]: let **𝛁discriminator** = TensorFlow.gradient(at: **discriminator**) { **discriminaotr** -> Tensor<Float> in

File: Sources/TensorFlow/Optimizers/MomentumBased.swift (hunk 1 above)
Comment (8bitmp3) [reply Sources/TensorFlow/Optimizers/MomentumBased.swift]: Thanks @t-ae , was writing late night/early morning

**Comment by 8bitmp3** [comment]
cheers

image

**Pull request merged**